### DISCUSSION TITLE
is sqlite db created by the Todo sql scripts ?

### DISCUSSION THREAD

**Discussion by TERIIHUB** [sqlpage/SQLPage#202] [General]
Hi everyone,
trying to build my first sqlweb site from : https://sql.ophir.dev/your-first-sql-website/
I've the following msg from the webpage after pressing "submit" : 
```
Backtrace
error returned from database: (code: 1) no such table: todo
```
my question is : does one have to create an sqlite table previous to execute the sqlpage server ..or do one of the sql script take care of that ?

PS : any other sqlpage beginner tutorials with step by step doc ?

Thanks

**Comment by lovasoa** [comment]
The tutorial at https://sql.ophir.dev/your-first-sql-website/ does not mention a `todo` table or even the word `todo`.

Can you share your exact code so that we can help you with it ?

**Comment by lovasoa** [comment]
Oh, I see, you are not talking about the tutorial at https://sql.ophir.dev/your-first-sql-website/ , but about the online example repl at https://replit.com/@pimaj62145/SQLPage#index.sql , which is linked from the tutorial.

It looks like you copied a single file (`index.sql`) locally, so the project does not work on your computer. You can use and modify the online example directly on repl.it, but if you want to duplicate locally on your computer, you will have to copy it in its entirety (not just the `index.sql` file).

You can use the **download as zip** feature from repl.it to do this in a single click : 

![image](https://github.com/lovasoa/SQLpage/assets/552629/7df31ca2-3842-4fba-ba0f-19a45037476d)

**Reply by lovasoa** [reply]
I'll answer the more general question from the title of this discussion, too. So, **where and how is the initial database created and populated?**

The default database connection string in SQLPage ([`DATABASE_URL`](https://github.com/lovasoa/SQLpage/blob/main/configuration.md)) is `sqlite://sqlpage.db?mode=rwc`. `rwc` stands for *read, write, create*. This means that SQLPage will try to open a sqlite database file named `sqlpage.db` in the current working directory, and, if it doesn't exist, will create a new empty database at this location. 

Once the database connection is opened, regardless of the database, SQLPage will look for sql migration files in the `sqlpage/migrations` directory. If it finds any, it will execute the ones that haven't been executed before, in order, allowing you to control the creation and modification of database tables when SQLPage starts. 

For instance, the replit example you mentioned contains a file named `sqlpage/migrations/0001_create_users_table.sql` which contains the following:

```sql
CREATE TABLE todo (
    id INTEGER PRIMARY KEY AUTOINCREMENT,
    todo_title TEXT NOT NULL
);
```

**Reply by TERIIHUB** [reply]
oh, okay, thanks a lot for those details ophir.
I will try your advice - in particular the zip downloading from the online Repl ! thanks a lot !
